libraries/AC_Avoidance/AC_Avoid.cpp

```diff
@@ -353,12 +403,70 @@ void AC_Avoid::limit_velocity(float kP, float accel_cmss, Vector2f &desired_vel_
 }
 
 /*
- * Compute the back away velocity required to avoid breaching margin
+ * Note: This method is used to limit velocity horizontally and vertically given a 3D desired velocity vector 
+ * Limits the component of desired_vel_cms in the direction of the obstacle_vector based on the passed value of "margin"
+ */
+void AC_Avoid::limit_velocity_3D(float kP, float accel_cmss, Vector3f &desired_vel_cms, const Vector3f& obstacle_vector, float margin_cm, float kP_z, float accel_cmss_z, float dt)
+{  
+    if (desired_vel_cms.is_zero()) {
+        // nothing to limit
+        return;
+    }
+    // create a margin_cm length vector in the direction of desired_vel_cms
+    // this will create larger margin towards the direction vehicle is traveling in
+    const Vector3f margin_vector = desired_vel_cms.normalized() * margin_cm;
+    const Vector2f limit_direction_xy{obstacle_vector.x, obstacle_vector.y};
+    
+    if (!limit_direction_xy.is_zero()) {
+        const float distance_from_fence_xy = MAX((limit_direction_xy.length() - Vector2f{margin_vector.x, margin_vector.y}.length()), 0.0f);
+        Vector2f velocity_xy{desired_vel_cms.x, desired_vel_cms.y};
+        limit_velocity_2D(kP, accel_cmss, velocity_xy, limit_direction_xy.normalized(), distance_from_fence_xy, dt);
+        desired_vel_cms.x = velocity_xy.x;
+        desired_vel_cms.y = velocity_xy.y;
+    }
+    
+    if (is_zero(desired_vel_cms.z)) {
+        // nothing to limit vertically 
+        return;
+    }
+
+    if (is_negative(desired_vel_cms.z * obstacle_vector.z)) {
```

Comment:
this is ok but a more obvious check might be, "if (is_positive(desired_vel_cmx.z) != is_positive(obstacle_vector.z))".  We need to be slightly careful about handling of zero because "is_positive()" will return true for 0.